cmd/archive_upload_mass_upload/mass_upload.go

```diff
@@ -0,0 +1,287 @@
+}
+
+func connectFtp(site string) (*ftp.ServerConn, error) {
+	conn, err := ftp.Dial(site, ftp.DialWithTimeout(dialTimeout))
+	return conn, err
+}
+
+// newGRPC makes a new grpc (over https) connection for the upload service.
+// host is the hostname to connect to, saPath is a path to a stored json service account key.
+func newGRPC(ctx context.Context, host, saPath string) (*grpc.ClientConn, error) {
+	var opts []grpc.DialOption
+
+	var idTokenSource oauth2.TokenSource
+	var err error
+	audience := "https://" + strings.Split(host, ":")[0]
+	if saPath == "" {
+		idTokenSource, err = idtoken.NewTokenSource(ctx, audience)
+		if err != nil {
+			if err.Error() != `idtoken: credential must be service_account, found "authorized_user"` {
+				return nil, fmt.Errorf("idtoken.NewTokenSource: %v", err)
+			}
+			gts, err := google.DefaultTokenSource(ctx)
+			if err != nil {
+				return nil, fmt.Errorf("attempt to use Application Default Credentials failed: %v", err)
+			}
+			idTokenSource = gts
+		}
+	} else {
+		idTokenSource, err = idtoken.NewTokenSource(ctx, audience, idtoken.WithCredentialsFile(saPath))
+		if err != nil {
+			return nil, fmt.Errorf("unable to create TokenSource: %v", err)
+		}
+	}
+
+	opts = append(opts, grpc.WithAuthority(host))
+
+	systemRoots, err := x509.SystemCertPool()
+	if err != nil {
+		return nil, err
+	}
+	cred := credentials.NewTLS(&tls.Config{
+		RootCAs: systemRoots,
+	})
+
+	opts = append(opts,
+		[]grpc.DialOption{
+			grpc.WithTransportCredentials(cred),
+			grpc.WithDefaultCallOptions(grpc.MaxCallSendMsgSize(maxMsgSize)),
+			grpc.WithPerRPCCredentials(oauth.TokenSource{idTokenSource}),
+		}...,
+	)
+
+	return grpc.Dial(host, opts...)
+}
+
+func new(ctx context.Context, aUser, aPasswd, site, bucket, grpcService, svcAccountKey string) (*client, error) {
+	f, err := connectFtp(site)
+	if err != nil {
+		return nil, fmt.Errorf("failed to connect to the ftp site(%v): %v", site, err)
+	}
+
+	// Login to cloud-storage, and get a bucket handle to the archive bucket.
+	if err := f.Login(aUser, aPasswd); err != nil {
+		return nil, fmt.Errorf("failed to login to site(%v) as u/p (%v/%v): %v",
+			site, aUser, aPasswd, err)
+	}
+
+	c, errS := storage.NewClient(ctx)
+	if errS != nil {
+		return nil, fmt.Errorf("failed to create a new storage client: %v", errS)
+	}
+	// Get a BucketHandle, which enables access to the objects/etc.
+	bh := c.Bucket(bucket)
+
+	// Create a new upload service client.
+	gc, err := newGRPC(ctx, grpcService, svcAccountKey)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create gRPC client: %v", err)
+	}
+
+	return &client{
+		gClient: pb.NewRVClient(gc),
+		bs:      c,
+		bh:      bh,
+		fc:      f,
+		bucket:  bucket,
+		ch:      make(chan *evalFile),
+	}, nil
+}
+
+// close politely closes the handles to cloud-storage and the ftp archive.
+func (c *client) close() {
+	c.fc.Quit()
+	if err := c.bs.Close(); err != nil {
+		glog.Fatalf("failed to close the cloud-storage client: %v", err)
```

Comment:
glog is essentially what we use internally...
I prefer it because it put things where I expect :) but...
I'm not really wedded to it :) It's handy that logs end up in (for a single run):
  /tmp/mo.donkey.morrowc.log.INFO.20220119-162131.1670723

and that /tmp/mo.INFO points to the latest file.. for instance.